# PyQt_GUI/stack.py

from PyQt5 import QtWidgets, uic
import sys, os

# ê²½ë¡œ ì„¤ì •
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# Controller import
from PyQt_Service.Monitoring.monitoring_controller import MonitoringController
from PyQt_Service.Setting.setting_controller import SettingController
from PyQt_Service.Dashboard.dashboard_controller import DashboardController
from PyQt_Service.Log.log_controller import LogController
from PyQt_Service.Log.log_manager import LogManager
from PyQt_Service.Monitoring.monitoring_collector import MonitoringCollector


class StackApp(QtWidgets.QMainWindow):
    def __init__(self):
        super().__init__()

        base_path = os.path.dirname(os.path.abspath(__file__))

        # main UI ë¡œë“œ (ëŒ€ì‹œë³´ë“œ ë©”ì¸ ìœˆë„ìš°)
        uic.loadUi(os.path.join(base_path, "dashboard.ui"), self)
        self.stack: QtWidgets.QStackedWidget = self.findChild(
            QtWidgets.QStackedWidget, "stackedWidget"
        )

        # ê° í˜ì´ì§€ ë¡œë“œ
        dashboard_page = self.stack.widget(0)  # ëŒ€ì‹œë³´ë“œ ì²« í˜ì´ì§€
        page_sungp = uic.loadUi(os.path.join(base_path, "sungp.ui"))
        page_log = uic.loadUi(os.path.join(base_path, "log.ui"))
        page_setting = uic.loadUi(os.path.join(base_path, "new_set.ui"))
        page_info = uic.loadUi(os.path.join(base_path, "information.ui"))

        # stackedWidget ì— í˜ì´ì§€ ë“±ë¡
        self.stack.addWidget(page_sungp)
        self.stack.addWidget(page_log)
        self.stack.addWidget(page_setting)
        self.stack.addWidget(page_info)

        # ì‹œìŠ¤í…œ ìƒíƒœ (ëŒ€ì‹œë³´ë“œ + ì„¤ì • í˜ì´ì§€ì—ì„œ ê³µìœ )
        self.system_state = {
            "pilot": "RED",          # ê¸°ë³¸ RED
            "halogen": False,
            "fan_commercial": False,
            "fan_battery": False,
        }

        # ğŸ”¹ ë¡œê·¸ ì»¨íŠ¸ë¡¤ëŸ¬ ìƒì„± + LogManager ì— ë“±ë¡
        self.log_controller = LogController(page_log)
        LogManager.instance().set_controller(self.log_controller)

        # ğŸ”¹ ëª¨ë‹ˆí„°ë§ ì»¨íŠ¸ë¡¤ëŸ¬
        self.monitoring_controller = MonitoringController(page_sungp)
        # CSV ë³´ê¸° ë²„íŠ¼
        page_sungp.btn_show_csv.clicked.connect(
            self.monitoring_controller.show_csv_table
        )

        # ğŸ”¹ ì„¤ì • ì»¨íŠ¸ë¡¤ëŸ¬
        self.setting_controller = SettingController(page_setting, self.system_state)

        self.collector = MonitoringCollector(self.setting_controller.serial)
        self.collector.start()

        # ğŸ”¹ ëŒ€ì‹œë³´ë“œ ì»¨íŠ¸ë¡¤ëŸ¬
        self.dashboard_controller = DashboardController(
            dashboard_page,
            self.setting_controller.serial,
            self.setting_controller.system_state,
        )
        
        self.setting_controller.dashboard = self.dashboard_controller

        # í˜ì´ì§€ ë¦¬ìŠ¤íŠ¸ (ë²„íŠ¼ ìˆœì„œì™€ ë§¤ì¹­)
        self.pages = [dashboard_page, page_sungp, page_log, page_setting, page_info]

        # ì™¼ìª½ ë©”ë‰´ ë²„íŠ¼ë“¤
        self.buttons = [
            self.pushButton,  # ëŒ€ì‹œë³´ë“œ
            self.pushButton_2,  # íƒœì–‘ê´‘ M
            self.pushButton_3,  # ë¡œê·¸
            self.pushButton_4,  # ì„¤ì •
            self.pushButton_5,  # ì •ë³´
        ]

        # ë²„íŠ¼ í´ë¦­ ì‹œ í˜ì´ì§€ ë³€ê²½
        for btn, page in zip(self.buttons, self.pages):
            btn.clicked.connect(lambda _, p=page: self.change_page(p))

        # ì¢…ë£Œ ë²„íŠ¼
        self.pushButton_6.clicked.connect(self.close)
        self.pushButton_6.setStyleSheet(
            "text-align: left; padding-left: 10px; color: #333333;"
        )

        # ì´ˆê¸° í˜ì´ì§€: ëŒ€ì‹œë³´ë“œ
        self.change_page(self.pages[0])

    def change_page(self, page):
        self.stack.setCurrentWidget(page)
        idx = self.pages.index(page)
        for i, btn in enumerate(self.buttons):
            style = btn.styleSheet()
            style = remove_color_from_stylesheet(style)
            if i == idx:
                btn.setStyleSheet(style + "color: #9E1010;")
            else:
                btn.setStyleSheet(style + "color: #333333;")

    def closeEvent(self, event):
        """
        í”„ë¡œê·¸ë¨ ì¢…ë£Œ ì‹œ Serial í¬íŠ¸ê°€ ì—´ë ¤ ìˆë‹¤ë©´ ìë™ìœ¼ë¡œ ë‹«ì•„ì£¼ëŠ” í•¨ìˆ˜.
        í¬íŠ¸ê°€ ë‹«íˆì§€ ì•Šìœ¼ë©´ ìš´ì˜ì²´ì œê°€ í¬íŠ¸ë¥¼ ê³„ì† ì ìœ í•˜ì—¬
        ë‹¤ìŒ ì‹¤í–‰ì—ì„œ ì—°ê²°ì´ ì•ˆ ë˜ëŠ” ë¬¸ì œë¥¼ ë°©ì§€í•œë‹¤.
        """
        try:
            if hasattr(self, "setting_controller"):
                serial = self.setting_controller.serial
                if serial.port and serial.port.is_open:
                    serial.port.close()
                    print("ğŸ”Œ ì‹œë¦¬ì–¼ í¬íŠ¸ ì •ìƒ ì¢…ë£Œë¨")
        except Exception as e:
            print(f"âš ï¸ ì‹œë¦¬ì–¼ í¬íŠ¸ ì¢…ë£Œ ì¤‘ ì˜¤ë¥˜: {e}")

        event.accept()


def remove_color_from_stylesheet(style):
    import re

    return re.sub(r"color\s*:\s*#[0-9A-Fa-f]+;", "", style)


if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)
    window = StackApp()
    window.show()
    sys.exit(app.exec_())
